### PR TITLE
Root cause use case scripts 

### LINKED ISSUE #453: [FEA]: Root cause analysis
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem this feature solves

This use case will help Morpheus to detect unknown root causes of failures using NLP

### Describe your ideal solution

It will use a binary sequence classifier to learn the known error types so that it can detect similar but previously unknown errors

### Describe any alternatives you have considered

_No response_

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by gbatmaz**
Raising this PR for root cause use case. Data is pending approval.

Depends on  #470 (Root cause analysis datasets]
Closes #453

### CODE REVIEW

**Review by raykallen — CHANGES_REQUESTED** [review]
Notebook mentions CLX but does not use CLX.

File: models/README.md [hunk 1]
```diff
@@ -165,3 +165,20 @@ For each process_id and snapshot there is a probablity score between 1 and 0, wh
...
+
+## Root Cause Analysis
+### Model Overview
+Root cause analysis is a binary classifier differentiating between ordinary logs and errors/problems/root causes in the log files.
+### Model Architecture
+BERT-base uncased transformer model
+### Training
+Training consisted of fine-tuning the original pretrained [model from google](https://huggingface.co/bert-base-uncased). The labeled dataset is linux kernel logs and it has two parts. the Unseen error logs will be appended to the testset
+### How To Use This Model
+This model is an example of customized transformer-based root cause analysis. It can be further fine-tuned for specific root cause analysis or predictive maintenance needs and of your enterprise using the fine-tuning scripts in the repo. The hyper parameters can be optimised to adjust to get the best results with your dataset. The aim is to get the model to predict some false positives that could be previously unknown error types.
```
[the remark this answers is not in the document]
Comment (gbatmaz) [reply models/README.md]: Thank you for this addition. Just added that part.

File: models/model-information.csv [hunk 2]
```diff
@@ -1,8 +1,9 @@
...
...nary classifier differentiating between rootcause and ordinary logs.,bert-base-uncased transformer model,Root cause analysis is a binary classifier differentiating between ordinary logs and errors/problems/root causes in the log files,This model is an example of customized transformer-based root cause analysis. It can be further fine-tuned for specific root cause analysis or predictive maintenance needs and of your enterprise using the fine-tuning scripts in the repo. The hyper parameters can be optimised to adjust to get the best results with your dataset. The aim is to get the model to predict some false positives that could be previously unknown error types.,Linux Kernel logs,Binary sequence classification as Root cause or ordinary logs,This model was trained on a very small sample subset it may not perform with other types of data without finetuning,N/A,https://arxiv.org/abs/1810.04805,1,128,V100,0.94 Higher accuracy may not always indicate success for this use case,0.96 Higher F1 may not always indicate success for this use case,0.94 Higher accuracy may not always indicate success for this use case,417MB,N/A,bert-base-uncased,128,N/A,TRUE,TRUE,11,1.7.1,3.8.10,18.04.5 LTS,4.6.1
```
[the remark this answers is not in the document]
Comment (gbatmaz) [reply models/model-information.csv]: I just fixed this

**Comment by gbatmaz** [comment]
> Notebook mentions CLX but does not use CLX.

Just fixed this in the branch

**Comment by gbatmaz** [comment]
> @gbatmaz @raykallen are we not going to include multi-class?

I think we should leave MC off at the moment. At least for this particular example.

File: models/model-information.csv [hunk 3]
```diff
@@ -1,8 +1,9 @@
...
```
Comment (efajardo-nv) [inline models/model-information.csv:9]: root cause model name needs to be updated here.

File: models/validation-inference-scripts/root-cause-models/root-cause-inference.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+# limitations under the License.
+
+"""
+Example Usage:
+python \
+root-cause-inference.py \
+    --validationdata ../../datasets/validation-data\
+    /root-cause-validation-data.jsonlines \
+    --model ../../root-cause-models/rootcause-binary-bert-20221118.onnx \
+    --vocab ../../training-tuning-scripts/rootcause-models\
```
Comment (efajardo-nv) [inline models/validation-inference-scripts/root-cause-models/root-cause-inference.py:23]: update path to root-cause-models

File: models/model-information.csv (hunk 3 above)
Comment (gbatmaz) [reply models/model-information.csv]: changed with the latest name

File: models/training-tuning-scripts/root-cause-models/sequence_classifier.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        --------
+        >>> sc.init_model("bert-base-uncased")  # huggingface pre-trained model
+        >>> sc.load_checkpoint(PATH)
+        """
+
+        model_dict = torch.load(file_path)
+        self._model.module.load_state_dict(model_dict["state_dict"])
+
+    def _get_hash_table_path(self):
+        hash_table_path = "%s/resources/bert-base-uncased-hash.txt" % os.path.dirname(os.path.realpath(__file__))
```
Comment (efajardo-nv) [inline models/training-tuning-scripts/root-cause-models/sequence_classifier.py:221]: this path is invalid within morpheus repo

File: models/training-tuning-scripts/root-cause-models/sequence_classifier.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        self._model.train()  # Enable training mode
+        self._tokenizer = SubwordTokenizer(self._hashpath, do_lower_case=True)
+
+        for _ in trange(epochs, desc="Epoch"):
+            tr_loss = 0  # Tracking variables
+            nb_tr_examples, nb_tr_steps = 0, 0
+            for df in train_dataloader.get_chunks():
+                b_input_ids, b_input_mask = self._bert_uncased_tokenize(df["text"], max_seq_len)
+
+                b_labels = torch.tensor(df["label"].to_array())
```
Comment (efajardo-nv) [inline models/training-tuning-scripts/root-cause-models/sequence_classifier.py:103]: `to_array()` is no longer in cudf 22.08. should `to_numpy()`

File: models/training-tuning-scripts/root-cause-models/sequence_classifier.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        test_gdf["label"] = labels
+
+        test_dataset = Dataset(test_gdf)
+        test_dataloader = DataLoader(test_dataset, batchsize=batch_size)
+
+        eval_accuracy = 0
+        nb_eval_steps = 0
+        for df in test_dataloader.get_chunks():
+            b_input_ids, b_input_mask = self._bert_uncased_tokenize(df["text"], max_seq_len)
+            b_labels = torch.tensor(df["label"].to_array())
```
Comment (efajardo-nv) [inline models/training-tuning-scripts/root-cause-models/sequence_classifier.py:151]: to_array() is no longer in cudf 22.08. should to_numpy()

File: models/validation-inference-scripts/root-cause-models/root-cause-inference.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    /resources/bert-base-uncased-hash.txt \
```
Comment (efajardo-nv) [inline models/validation-inference-scripts/root-cause-models/root-cause-inference.py:24]: not a valid `vocab` path. `resources` directory does not exist. should probably point to `morpheus/data/bert-base-uncased-hash.txt`

File: models/training-tuning-scripts/root-cause-models/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,8 @@
+cudf==0+untagged.1.g788bddd
```
Comment (efajardo-nv) [inline models/training-tuning-scripts/root-cause-models/requirements.txt:1]: should this be same cudf version used by morpheus?

File: models/training-tuning-scripts/root-cause-models/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+cupy==9.1.0
+numpy==1.23.4
+onnxruntime==1.13.1
+scipy==1.6.0
+torch==1.7.1
```
Comment (efajardo-nv) [inline models/training-tuning-scripts/root-cause-models/requirements.txt:6]: morpheus uses torch 1.10.2

File: models/validation-inference-scripts/root-cause-models/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,7 @@
+cudf==0+untagged.1.g788bddd
```
Comment (efajardo-nv) [inline models/validation-inference-scripts/root-cause-models/requirements.txt:1]: should this be same cudf version used by morpheus?

File: models/validation-inference-scripts/root-cause-models/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,7 @@
...
+cupy==9.1.0
+numpy==1.23.4
+onnxruntime==1.13.1
+scipy==1.6.0
+torch==1.7.1
```
Comment (efajardo-nv) [inline models/validation-inference-scripts/root-cause-models/requirements.txt:6]: morpheus uses torch 1.10.2. should/can this be updated to same?

File: models/training-tuning-scripts/root-cause-models/root-cause-bert.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Example Usage:
+python root-cause-bert.py \
+--trainingdata ../../datasets/training-data/root-cause-training-data.csv
+--unseenerrors ../../datasets/training-data/root-cause-unseen-errors.csv
```
Comment (efajardo-nv) [inline models/training-tuning-scripts/root-cause-models/root-cause-bert.py:19]: need end of line back slashes

File: models/validation-inference-scripts/root-cause-models/root-cause-inference.py (hunk 8 above)
Comment (gbatmaz) [reply models/validation-inference-scripts/root-cause-models/root-cause-inference.py:24]: I created that path now added the vocab file

Comment (gbatmaz) [reply models/validation-inference-scripts/root-cause-models/root-cause-inference.py:24]: which seems to be the same way as phishing.

File: models/training-tuning-scripts/root-cause-models/sequence_classifier.py (hunk 5 above)
Comment (gbatmaz) [reply models/training-tuning-scripts/root-cause-models/sequence_classifier.py:221]: I've created that file and folder which we also had for phishing detection. so they follow the same convention now

File: models/training-tuning-scripts/root-cause-models/root-cause-bert.py (hunk 13 above)
Comment (gbatmaz) [reply models/training-tuning-scripts/root-cause-models/root-cause-bert.py]: done

File: models/validation-inference-scripts/root-cause-models/requirements.txt (hunk 12 above)
Comment (gbatmaz) [reply models/validation-inference-scripts/root-cause-models/requirements.txt]: updated the torch version.

File: models/validation-inference-scripts/root-cause-models/requirements.txt (hunk 11 above)
Comment (gbatmaz) [reply models/validation-inference-scripts/root-cause-models/requirements.txt]: upgraded the versions

File: models/training-tuning-scripts/root-cause-models/sequence_classifier.py (hunk 7 above)
Comment (gbatmaz) [reply models/training-tuning-scripts/root-cause-models/sequence_classifier.py]: updated the scripts

File: models/training-tuning-scripts/root-cause-models/sequence_classifier.py (hunk 6 above)
Comment (gbatmaz) [reply models/training-tuning-scripts/root-cause-models/sequence_classifier.py]: updated for later versions of cudf

File: models/training-tuning-scripts/root-cause-models/requirements.txt (hunk 10 above)
Comment (gbatmaz) [reply models/training-tuning-scripts/root-cause-models/requirements.txt]: changed the req, according to the updated env

File: models/training-tuning-scripts/root-cause-models/requirements.txt (hunk 9 above)
Comment (gbatmaz) [reply models/training-tuning-scripts/root-cause-models/requirements.txt]: upgraded the versions

**Review by raykallen — APPROVED** [review]

**Comment by raykallen** [comment]
@gpucibot merge
